### ISSUE TITLE
Question about the loss

### ISSUE THREAD

**Issue by soon-will** [jiangqy/DPSH-pytorch#2]
Thanks for your effort, but I have two question about the loss:

1. In the ` Logtrick` function, why do you add `torch.max(x, Variable(torch.FloatTensor([0.])))`;

2. In the end, why do you divide `num_train * len(train_label)`.

Thank you for your reply.

**Comment by jiangqy** [comment]
@soon-will 
1. we have the following equation: log(1 + e^x) = log(1 + e^{-|x|}) + max(0, x). In practice, the former suffers from numerical problem. For example:
![image](https://user-images.githubusercontent.com/14566109/49194171-ff741900-f3bc-11e8-860d-d1d0ee6ba5d6.png)

2. Please note that you can discard this term, e.g., by discarding this term, set lr = lr / (num_train * len(train_label)), you should get the same result. However, different dataset has different num_train * len(train_label). We add this term just for easy adjustment for learning rate.

**Comment by soon-will** [comment]
@jiangqy Thank you for your reply.

**Comment by soon-will** [comment]
@jiangqy  Sorry to trouble you, but I still have some trouble about the loss.

![image](https://user-images.githubusercontent.com/25767468/49350609-8c371380-f6ea-11e8-9cd1-6c610110d373.png)

For the first term of the loss, I think the term is divided by `num_train * len(train_label)` to average the loss since there are `num_train * len(train_label)` pairs of codes(`num_train` is the number of the train set, and `len(train_label)` is the number of the batch size, as the shape of `S` and `theta_x` is (batch_size, num_train)). 

For the second term of the loss,  I think the term should be divided by `batch_size` to average the term since there are `batch_size` pairs of codes. But the code you provided is also divided by `num_train * len(train_label)`.

By the way, could I send you a email when have trouble?

Thank you.

**Comment by jiangqy** [comment]
@soon-will 
It's okay if you want dividing term as the way you describe. The difference is just the value of $\eta$ is different.

my email is jiangqy#lamda.nju.edu.cn

**Comment by soon-will** [comment]
Thank you for your reply.
